Implementation of units in models [simplified version of #4615]

astrofrog opened the pull request:
_This is a simplified version of https://github.com/astropy/astropy/pull/4615 - in particular it no longer includes all the unit spec implementation, and I've tried to remove any changes unrelated to the functionality being added here_

**Feedback needed!** it would be great if people could try and clone this branch and try out various cases with the `Gaussian1D` model. In addition, I would really appreciate code reviews!

**NOTE:** please read my [latest comment](https://github.com/astropy/astropy/pull/4855#issuecomment-278408203) before reviewing this!

This pull request finishes the basic implementation of support for units in models and fitting. I'll try and give some background to the main changes here. At the moment, I've set units as opt-in on a per-model basis, and only the Gaussian1D model supports units for now.
### Units on parameters and unit-friendly evaluation

For models that support units, parameters can now be quantities:

```
>>> from astropy.modeling.models import Gaussian1D
>>> g = Gaussian1D(amplitude=1 * u.J, mean=1 * u.m, stddev=0.1 * u.m)
```

Support for units of parameters and evaluation of models with parameters that have units can be enhanced by adding a propery called `input_units` that returns the units that the input should be in (and `None` can be used to indicate that any unit can be passed in). For `Gaussian1D`, this is:

``` python
    @property
    def input_units(self):
        if self.mean.unit is None:
            return {'x': dimensionless_unscaled}
        else:
            return {'x': self.mean.unit}
```

This allows us to check that the input has the correct units:

```
>>> g(1)
...
UnitsError: Units of input 'x', (dimensionless), could not be converted to required input units of m (length)

>>> g(3 * u.cm)
Out[13]: <Quantity 3.703531977652397e-21 J>
```

Note that `input_units` is just a convenience that allows us to give clearer error messages, but even if it is not present, quantities will still be passed to the evaluation function which is free to raise its own errors (this was a point that @hamogu and @mhvk made in https://github.com/astropy/astropy/pull/4615#discussion_r53538398).

### Equivalencies on parameters

Equivalencies on parameters are deliberately **not supported**. This is because it would introduce conceptual difficulties: a Gaussian in wavelength space is not the same as a Gaussian in frequency space.

### Equivalencies on input

However, one place where equivalencies can be unambiguous is when **evaluating** the model, for the input data values. These equivalencies are passed through a new keyword argument `equivalencies` in the model call:

```
>>> g(3 * u.Hz, equivalencies={'x': u.spectral()})
<Quantity 0.0 J>
```

This requires the ``input_units`` property to exist and converts the input values to the ``input_units`` units.

### Fitting models/data with units

Adding support for fitting models to data with units is done by adding a method named `_parameter_units_for_data_units` (for lack of a better name) that takes the units of the data and returns the units that each parameter would have to make sure all the units were consistent. For `Gaussian1D`, this looks like:

``` python
    def _parameter_units_for_data_units(self, xunit, yunit, zunit):
        return OrderedDict([('mean', xunit),
                            ('stddev', xunit),
                            ('amplitude', yunit)])
```

In order to implement support for models and/or data with units for a particular fitter, one simply needs to add the `fitter_unit_support` decorator to the `__call__` function. With that in place, we can do things like:

``` python
>>> import numpy as np
>>> from astropy import units as u
>>> from astropy.modeling import models, fitting

>>> x = np.linspace(-5., 5., 2000)
>>> y = 3 * np.exp(-0.5 * (x - 1.3)**2 / 0.8**2)
>>> y += np.random.normal(0., 0.2, x.shape)
>>> x = x * u.m
>>> y = y * u.Jy

>>> g_init = models.Gaussian1D(amplitude=1. * u.mJy, mean=3 * u.cm, stddev=2 * u.mm)
>>> fit_g = fitting.LevMarLSQFitter()
>>> g = fit_g(g_init, x, y)

>>> print(g)
Model: Gaussian1D
Inputs: ('x',)
Outputs: ('y',)
Model set size: 1
Parameters:
     amplitude        mean         stddev    
         Jy            m             m       
    ------------ ------------- --------------
    3.0161309672 1.30493337628 0.797252253802
```

Behind the scenes, this does the following:
1. Determine the units the parameters would need to be in so that if we stripped the units from the data and model, everything would be correct relative to each other.
2. Create a copy of the model, then convert the values to the units determined in the previous step, then drop the units, and drop the units on the data too
3. Do the fitting (with no units)
4. Add back the units determined in step 1 to the model copy

pllim commented on astropy/modeling/core.py:
```diff
@@ -2871,10 +3024,7 @@ def _prepare_outputs_single_model(model, outputs, format_info):
-        return outputs[0]
-    else:
-        return tuple(outputs)
+    return tuple(outputs)
```
What is the reasoning of this change? It seems unrelated to unit support at a glance.

astrofrog replied:
This was one of the original changes by @embray (I forgot to revert that for now and put it in a separate PR)

pllim replied:
Ah, okay. Now I see that this logic is moved further up.

pllim commented:
Some comments based on your examples:

1. When input unit is defined and I type `g(1)`, I would expect the 1 to be treated as having `g.mean.unit` by default without me having to override `input_units` property. Or is it just me?
2. When I set up the model as `Gaussian1D(amplitude=1, mean=5000 * u.AA, stddev=100)`, it did not complain that `mean` was assigned an unit, but not `stddev`, which might lead an user to believe that `stddev` is automatically given `g.mean.unit`. However, this is clearly not the case because I get `TypeError` on evaluation.
3. `Linear1D(3 * u.m)` did not raise the error as you indicated in your example above.

pllim commented:
Also, this breaks the `bounding_box` property:
```python
>>> g.bounding_box
.../astropy/modeling/core.pyc in bounding_box(self)
   1037             # with any optional parameters)
   1038             # (In other words, in this case self._bounding_box is a *class*)
-> 1039             bounding_box = self._bounding_box((), _model=self)()
   1040             return self._bounding_box(bounding_box, _model=self)
   1041 

.../astropy/modeling/core.py in __call__(self, factor)
    354             kwargs.append((param.name, param.default))
    355 
--> 356         __call__ = make_function_with_signature(__call__, ('self',), kwargs)
    357 
    358         return type(str('_{0}BoundingBox'.format(cls.name)), (_BoundingBox,),

.../astropy/modeling/core.pyc in __call__(self, **kwargs)
    336 
    337         def __call__(self, **kwargs):
--> 338             return func(self._model, **kwargs)
    339 
    340         kwargs = []

.../astropy/modeling/functional_models.pyc in bounding_box(self, factor)
    145         dx = factor * self.stddev
    146 
--> 147         return (x0 - dx, x0 + dx)
    148 
    149     @staticmethod

.../astropy/units/quantity.pyc in __array_prepare__(self, obj, context)
    343                                      "argument is not a quantity (unless the "
    344                                      "latter is all zero/infinity/nan)"
--> 345                                      .format(function.__name__))
    346             except TypeError:
    347                 # _can_have_arbitrary_unit failed: arg could not be compared

UnitsError: Can only apply 'subtract' function to dimensionless quantities when other argument is not a quantity (unless the latter is all zero/infinity/nan)
```

pllim commented:
Also, this breaks `n_models > 1`:
```python
>>> g = Gaussian1D([1 * u.J, 2.0], [1 * u.m, 5000 * u.AA], [0.1 * u.m, 100 * u.AA], n_models=2)
.../astropy/modeling/core.py in __init__(self, amplitude, mean, stddev, **kwargs)
    432 
    433             new_init = make_function_with_signature(
--> 434                     __init__, args, kwargs, varkwargs='kwargs')
    435             update_wrapper(new_init, cls)
    436             cls.__init__ = new_init

.../astropy/modeling/core.pyc in __init__(self, *params, **kwargs)
    429 
    430             def __init__(self, *params, **kwargs):
--> 431                 return super(cls, self).__init__(*params, **kwargs)
    432 
    433             new_init = make_function_with_signature(

.../astropy/modeling/core.pyc in __init__(self, *args, **kwargs)
    691         # Parameter values must be passed in as keyword arguments in order to
    692         # distinguish them
--> 693         self._initialize_parameters(args, kwargs)
    694 
    695     def __repr__(self):

.../astropy/modeling/core.pyc in _initialize_parameters(self, args, kwargs)
   1547                     continue
   1548                 else:
-> 1549                     params[param_name] = np.asanyarray(value, dtype=np.float)
   1550 
   1551         if kwargs:

.../numpy/core/numeric.pyc in asanyarray(a, dtype, order)
    523 
    524     """
--> 525     return array(a, dtype, copy=False, order=order, subok=True)
    526 
    527 def ascontiguousarray(a, dtype=None):

ValueError: setting an array element with a sequence.
```

When units are not involved, it works:
```python
>>> g = Gaussian1D([1, 2.0], [1, 5000], [0.1, 100], n_models=2)
>>> g
<Gaussian1D(amplitude=[ 1., 2.], mean=[    1., 5000.], stddev=[   0.1, 100. ], n_models=2)>
```

pllim commented:
Also... Once models can handle units, one of the first things I am going to do is to add a blackbody model (there was an attempt via #1480). So, it would be desirable to make sure that this PR can handle something like astrofrog/astropy#58.

mhvk commented:
@astrofrog - top level comment first: if we do equivalencies at all, the data will have to converted to the parameter's units rather than the other way around in order for a Gaussian to work (or the Gaussian has to turn itself in another function...), which has obvious problems with multiple models possibly requiring different transformations. I think it makes more sense to start by omitting equivalencies.

mhvk commented on astropy/modeling/core.py:
```diff
@@ -1068,6 +1080,49 @@ def has_user_bounding_box(self):
+
+        model = self.copy()
+
+        xunit = x.unit if isinstance(x, Quantity) else dimensionless_unscaled
```
Just `xunit = getattr(x, 'unit', dimensionless_unscaled)`?

astrofrog replied:
Fixed

mhvk commented:
Along the line of equivalencies: logarithmic units are going to be problematic here. Not a deal-breaker, but it makes me worry slightly about the approach. Possibly, it *is* better (even if much slower) to just let a model upon evaluation convert the data as appropriate for the parameters.

mhvk commented on astropy/modeling/core.py:
```diff
@@ -1068,6 +1080,49 @@ def has_user_bounding_box(self):
+        for name, unit in parameter_units.items():
+            parameter = getattr(model, name)
+            if parameter.unit is not None:
+                parameter.value *= parameter.unit.to(unit)
```
Not all transformations are multiplicative... Instead: `parameter.value = parameter.to(unit).value`.

mhvk commented on astropy/modeling/core.py:
```diff
@@ -1068,6 +1080,49 @@ def has_user_bounding_box(self):
+
+        for name, unit in parameter_units.items():
+            parameter = getattr(model, name)
+            parameter._set_unit(unit, force=True)
```
Shouldn't the value of the parameter be changed?

astrofrog replied:
This method is intended to be used with models that have parameters that have no units, so these lines are just adding units so there is no conversion going on.

mhvk commented:
Ai, this is not going to be a little bit of work... have to run...

astrofrog commented:
> Along the line of equivalencies: logarithmic units are going to be problematic here. Not a deal-breaker, but it makes me worry slightly about the approach. Possibly, it is better (even if much slower) to just let a model upon evaluation convert the data as appropriate for the parameters.

The issue is that the units of the data can't easily be passed to the data since the model is called from e.g. scipy optimizers which won't know to pass through the units. I think you're still going to have to wrap the fitters and modify the model in some way to let it know about the data units. I'm not convinced the current approach can't be extended to deal with log units - I will give it a try.

astrofrog commented:
@nden - as promised, this is now rebased, and ready for you to try out

MSeifert04 commented:
Seems like CircleCI doesn't like #5319 and the `posix.types` import.

astrofrog commented:
@MSeifert04 - see https://github.com/astropy/astropy/pull/5498

MSeifert04 commented:
👍

nden commented:
Thanks @astrofrog. I'll give it a try in the next few days.

pllim commented:
My concerns above still stand.

astrofrog commented:
@pllim - regarding your first three comments:

> When input unit is defined and I type g(1), I would expect the 1 to be treated as having g.mean.unit by default without me having to override input_units property. Or is it just me?

I don't think this would be very safe - sometimes you are not the one setting up the model - for instance if we have the blackbody function which requires wavelength and is defined in Astropy, you wouldn't want that function to be able to take unitless input for the evaluation. So I'd rather not have ``g(1)`` work in this case.

> When I set up the model as Gaussian1D(amplitude=1, mean=5000 * u.AA, stddev=100), it did not complain that mean was assigned an unit, but not stddev, which might lead an user to believe that stddev is automatically given g.mean.unit. However, this is clearly not the case because I get TypeError on evaluation.

The prior implementations of units in modeling included a lot of checking upon parameter setting which complicated the code base a lot. This PR tries to simplify this following the suggestion many people made that the consistency of units between parameters should be done at evaluation time as in your example here. One way to get around this without re-introducing a whole load of consistency checking between parameters would be to evaluate the model for one scalar value once any parameter is changed. There is a small performance penalty to this, so this could potentially be an option that could be turned off in the model init, for instance ``Gaussian1D(amplitude=1, mean=5000 * u.AA, stddev=100, check_parameter_units=False)``. What do you think?

> Linear1D(3 * u.m) did not raise the error as you indicated in your example above.

I'll investigate!

astrofrog commented:
@pllim - I've fixed the bounding box property for Gaussian1D in e39dd6fbffc7be28c2ec620e64499305f031dcfc when units are present. I will need to add a test for all the different models to make sure this works well since it is up to the bounding_box property (like evaluate) to support units.

I've also fixed the case for n_models > 1 with units in 6b4266c4bfc062acc495defcf0e1f6baee95c567.

astrofrog commented:
@pllim - regarding:

> Linear1D(3 * u.m) did not raise the error as you indicated in your example above.

Indeed this is not the case (because if units work in the evaluation, the model automatically supports units, so some models such as Linear1D 'just work'). So I've removed this from the PR description higher up.

astrofrog commented:
I've decided to remove equivalencies from parameters, because these introduce conceptual difficulties: a Gaussian in wavelength space is not a Gausssian in frequency space. So the model should be defined in a well defined unit system. However, it's still possible to specify equivalencies when **evaluating** the model, e.g. ``g(1 * u.Hz, equivalencies=u.spectral())``, which causes the input to be converted to ``input_units`` (as @mhvk was suggesting).

astrofrog commented:
@pllim @mhvk @MSeifert04 @nden @Cadair: This is ready for a penultimate round of review. Please take this for a spin. If people are ok with the approach here, I will go ahead and add documentation, tests, and a changelog entry, then it will be ready for final review, at which point I will email the list. We should make sure we get this in the next release! (it's long overdue...)

Cadair commented on astropy/modeling/core.py:
```diff
@@ -1179,6 +1250,38 @@ def prepare_inputs(self, *inputs, **kwargs):
 
+        # Check that the units are correct, if applicable
+
+        if self.input_units is not None:
```
Is this logic not very similar to that of `u.quantity_input`? If you made `input_units` return a dict it would become `**kwargs` to `quantity_input` and you could drop just use the decorator on the call to `evaluate` if `input_units` is not None?

Cadair commented:
Just to chime in on the non-existant unit checking at model creation, while I see @pllim's point about `stddev` and the error is not pretty there (under Python 3 it's 3 exceptions deep!), on the other hand this is quite nice:
```python
>>> g = m.Gaussian1D(amplitude=1000, mean=656.28*u.nm, stddev=0.2*u.nm)
>>> g(6562*u.AA)
923.11635
```
In that it happily performs unit validation on the input but leads to a unitless output. I am not saying that this is something we should support (I am all for unit's always everywhere), but I thought it to be an interesting side effect.

bernie-simon commented:
So here is the issue that's preventing me from creating a version of
BlackBody1D as a FittableModel. The black body model it is supposed to
wrap, in astropy.analytic_functions, exepcts either that the input
array is a Quanity or, if it is not, assumes the units. Since the
fittable model can get its input in any compatible uits, the
information about these units needs to be passed to the black body
function. Howver, before invoking this function the modeling code
Converts the input arrays from Quantities to unitless numpy arrays in
without_units_for_data. The information about the units is supposed to
be retained by the model parameters, but in the case of the balck body
model, the units of the parameter, temperature, has no direct
relationship to the units of the input arrays to the model.

I thought I could solve the problem with a little bit of
trickery. Every FittableModel has the method
_parameter_units_for_data_units, which gets passed the units of the
input arrays. I thought I could save the units as new attributes on
the model object:

    self.xunit = xunit
    self.yunit = yunit

However, there are two problems. First, the units are needed in the
evaluate method, but it is defined as a static method, and so self is
not passed to it. (This appears to be a mistake. The evaluate method
is called from __call__ in core.py as a normal object method and is
passed self. But the assumption that it is not a normal object method
is all throughout the code in functional_models.py.) Second, even
after the code is corrected so that evaluate is a normal object
method, the information saved in _parameter_units_for_data_units
cannot be passed to evaluate, as the object the new attributes are
saved on, model in fitter_unit_support, is thrown away and replaced by
another object, model_new, which was copied from the original model
before calling _parameter_units_for_data_units.

Neither of these problems would be hard to fix, but will need to be
addressed before making a black body fittable model. I hesitate making
them myself because I do not yet understand all the assumptions behind
the FittableModel code. Or perhaps some other way to access unit
information could be made available so I do not need to use the trick
described here.

Cadair commented:
hey @bernie-simon, I think I have at least part-solved this problem here: https://github.com/astrofrog/astropy/pull/65
I was doing that to handle the case where the evaluate function calls a function which is completely unit-unaware, but I think it could be easily used to address your strict unit-aware case.
Basically what I added, that I think will be useful, is a `_input_units_strict` attribute, which if set the model will convert the arguments (not the parameters) to `evaluate` to the units specified by either function annotations (python 3) or the `input_units` property.

I am not sure if I full understand the interaction with the unitless fitting stuff, but I think this would ensure that you always get the inputs to a certain unit. Does this make any sense?!

bernie-simon commented:
I'm not sure this would resolve the problem. Let me restate what I think the problem is. The input data to the model can have arbitrary units. The current code strips off the units from the input data and converts the model parameters to these units. This works for the case of fitting a gaussian, as the model parameters have units that are equivalent to the input parameters. But it fails in the case of the black body model. The units of the model parameter, temperature, are not equivalent to the input units. And the black body model needs access to the units of the input data to determine whether the units are in wavelength or frequency space.

I think the solution requires saving the information about the input units somewhere that can be queried by the model code. I tried to save the unit information myself, but that fails for reasons explained in my previous message. But I think it would be better to have a formal interface to do this. I do not think annotations on the code solves the problem, as the information the model needs is the units of the input data and not the units that the model would like to see.

Cadair commented:
Ah ok, I see, thanks for the extra explanation :)

astrofrog commented:
@bernie-simon @pllim - do you have a branch that has the code you have been working on? I'll take a loot at all this while traveling this evening.

astrofrog commented:
@bernie-simon @pllim - I had a chance to look at this on the :train: and I managed to get everything working. I incorporated @Cadair's changes, which allow models with units to also accept dimensionless input and add a unit to it on-the-fly, and I then also incorporated @pllim's commits to add the blackbody model. With this, everything 'works':

https://gist.github.com/astrofrog/816fa34c47e8f5996f55c96d5622d1fa

Note that I had to add an amplitude parameter, which I think should probably actually be a luminosity parameter. I need to sort this out and also figure out how to deal properly with the extra steradian unit. But as you can see, the evaluation and fitting work with the latest changes.

astrofrog commented:
Ah, I spoke too soon, this still needs to be adjusted to work with non-Hz input to the fitter, but this is fixable without much hacking.

astrofrog commented:
(still working on this, will explain how it works once it's all polished up)

astrofrog commented:
@bernie-simon @pllim - I've now documented the Blackbody1D class extensively to explain how I got this to work. I've updated the gist at: https://gist.github.com/astrofrog/816fa34c47e8f5996f55c96d5622d1fa

Cadair commented on astropy/modeling/functional_models.py:
```diff
@@ -2086,52 +2086,79 @@ class BlackBody1D(Fittable1DModel):
+    # in this case the input x values are assumed to be frequencies in Hz.
     input_units_allow_dimensionless = True
 
     def evaluate(self, x, temperature, amplitude):
```
amplitude should be `bolometric_flux`

astrofrog replied:
Oops, fixed

pllim commented on astropy/modeling/functional_models.py:
```diff
@@ -2062,3 +2078,87 @@ def evaluate(cls, x, y, amplitude, r_eff, n, x_0, y_0, ellip, theta):
+    # bolometric flux is the integral of the model over the spectral axis. This
+    # is more useful than simply having an amplitude parameter.
+    temperature = Parameter(default=5000 * u.K, min=0)
+    bolometric_flux = Parameter(default=1 * u.erg / u.cm ** 2 / u.s)
```
I don't get this one. Is "bolometric_flux" really necessary? What if I don't want any normalization?

My original idea is just to deprecate the blackbody functions in `analytic_functions` altogether (and in that process, get rid of that sub-package) and make this model do the calculations instead. The only reason we have `analytic_functions` at all was because we could not put blackbody in `modeling`. Now that we can, there is no reason to have duplicate functionality.

astrofrog replied:
Well if you want to actually fit the model to data, then you need to have some kind of 'vertical' scaling, so as a model it doesn't seem very useful without scaling. If you want the blackbody as is given by blackbody_nu, just don't change bolometric_flux (the default value is the right one to get the same values out). But many people will want to get the flux densities for a given bolometric flux/luminosity.

pllim replied:
Okay. But I think we should still add deprecation message in `analytic_functions` as part of this PR. There is no reason for `analytic_functions` to exist anymore.

astrofrog replied:
@pllim - I'd like to delay the decision to deprecate analytic_functions until after this PR, because there are currently discussions relating to how to deprecate things to minimize issues for users:

https://github.com/astropy/astropy-APEs/pull/17

I agree with you in principle though, but just don't think we need to rush.

pllim commented on astropy/modeling/parameters.py:
```diff
@@ -142,6 +170,10 @@ class Parameter(OrderedDescriptor):
+    unit : `~astropy.units.Unit`
+        if specified, the parameter will be in these units, and when the
+        parameter is updated in future, it should be set to a
+        :`~astropy.units.Quantity` that has equivalent units.
```
This looks like a typo. 
```
:`~astropy.units.Quantity`
```
should be
```
:class:`~astropy.units.Quantity`
```
?

pllim commented on astropy/modeling/core.py:
```diff
@@ -1363,7 +1511,8 @@ def _initialize_parameters(self, args, kwargs):
                 if value is None:
                     continue
-                params[param_name] = np.asanyarray(value, dtype=np.float)
+                else:
```
The `else` is unnecessary? I don't see you use it on Line 1500 above.

pllim commented on astropy/modeling/functional_models.py:
```diff
@@ -2062,3 +2078,87 @@ def evaluate(cls, x, y, amplitude, r_eff, n, x_0, y_0, ellip, theta):
+        """
+
+        from ..constants import sigma_sb
+        from ..analytic_functions.blackbody import blackbody_nu
```
If we agree to move ahead with deprecating `analytic_functions`, we should move the actual calculations here. That means, we also have to ensure `blackbody_lambda` can be done somehow. Also have to move over the tests from there.

I cannot seem to  get the equivalent of `test_blackbody_synphot()` to pass using this model. I'll have to investigate more tomorrow...

pllim commented on astropy/modeling/functional_models.py:
```diff
@@ -2062,3 +2078,87 @@ def evaluate(cls, x, y, amplitude, r_eff, n, x_0, y_0, ellip, theta):
+
+    Parameters
+    ----------
+    temperature : `~astropy.units.Quantity`
```
I am unable to initialize blackbody with a non-quantity, like other models like Gaussian. I propose that we allow using non-Quantity but just assume some default unit.

I also cannot pass in other temperature units that we support; i.e., Celsius and Fahrenheit.

astrofrog replied:
@pllim - I'm going to be working on this today

astrofrog commented:
Since we're now getting into lots of small details, I thought that before pushing any more changes I would make a notebook demonstrating the current behavior:

http://nbviewer.jupyter.org/gist/astrofrog/5d4760f3702ebd324855f85f43207bc0

@pllim @Cadair (and others) - could you take a look at this and let me know if you agree with the behavior shown in there? Let me know if there are things which you don't know whether they are possible or not and I'll add this to the notebook either way. Note that I'm leaving aside the specific example of the blackbody for now, let's come back to that once we're happy with how things work for the Gaussian.

pllim commented:
@astrofrog , I left some comments on your Gist page. Please let me know if you didn't get it.

embray commented:
Was just going back through some old e-mails (and I mean old, like from last May) and saw one that @astrofrog sent on this PR.  And now here I am and just as of last Friday the PR is still being discussed 🤣 

Is there anything I can do to help review this?  I can just barely remember what I did on #3852 (which IIRC got a good ways of the way there) but it seems like @astrofrog has added/changed a ton since then, for what I assume are good reasons.

astrofrog commented:
@embray - yes, there have been quite a few iterations on this... But I think we are almost there. If you have cycles to spare, could you review the gist I posted a link to in my previous comment to check if you have any comments about the behavior presented in there (note that not all changes have been pushed to this PR yet, so I'm only asking for feedback on the content of the gist).

bsipocz commented on astropy/modeling/functional_models.py:
```diff
@@ -2062,3 +2078,87 @@ def evaluate(cls, x, y, amplitude, r_eff, n, x_0, y_0, ellip, theta):
+        # calculate here. We convert to 1/Hz to make sure the units are
+        # simplified as much as possible, then we multiply by the bolometric
+        # flux to get the normalization right.
+        fnu = (np.pi * u.sr * blackbody_nu(x, temperature) /
```
I'm afraid that I've been lost in the details, and can't see where u.sr is coming from?

Cadair commented:
@astrofrog I suggest you push your changes used in the gist to here, I think it addresses most of the issues and I think you are more likely to get constructive feedback on the code (rather than the gist).

astrofrog commented:
Hi everyone 👋 

I think this is finally ready for a proper review (code, functionality, etc). There may still be some small details to figure out in behavior, but I think the main infrastructure is here. I've tried to comment many of the changes here to make them as clear as possible. For anyone who isn't comfortable looking at the modeling code itself, you can still look at the tests which should also be heavily commented.

In addition, I have made two notebooks demonstrating the functionality in this pull request:

Current general behavior: [notebook](https://gist.github.com/astrofrog/5d4760f3702ebd324855f85f43207bc0)

Using the blackbody model: [notebook](https://gist.github.com/astrofrog/816fa34c47e8f5996f55c96d5622d1fa)

I would really appreciate some careful reviews! :pray:

Note that not all models are yet unit-friendly. For instance, @nden found that the polynomial models need to have fixes in the evaluate method to properly work with units. So for now, focus on the Gaussian1D and Blackbody1D models, and I suggest that unit support for various models can be added in subsequent pull requests (think of this one as the one laying the foundation for it but not necessarily making absolutely everything work). We'll need to add tests for all models with units, so it's not going to be a small effort.

Another untested domain are compound models 😅 . If someone would like to try this out, please feel free to try it, but again I would like to defer this to future pull requests.

There are things that still need to be done in this pull request:

* [ ] Add more tests, especially for the blackbody fitting
* [ ] Add a changelog entry
* [ ] Updating the documentation

but I would like to do these only once people are happy with the rest of the pull request, since otherwise it's a lot of work to update the tests and docs and change them multiple times.

@pllim @nden - thank you for your previous review of the gist! I've incorporated some of the things here. I really appreciate the regular reviews!

pllim commented:
Ops. I didn't realize all the card moving stuff is visible here. Now my clumsiness with a mouse is recorded forever in GitHub history...

bsipocz commented:
I haven't reviewed the code but I very much like what I see in the blackbody gist.

pllim commented:
@astrofrog , I added more reviews to your Gists. Thanks for the hard work!

nocturnalastro commented:
_supports_unit_fitting doesn't seem to be preserved to compound models such as.
``` python
In [28]: g4 = Gaussian1D(mean=3 * u.micron, stddev=1 * u.micron, amplitude=1 * u.Jy)
In [29]: g44 = g4 + g4
In [30]: g4._supports_unit_fitting
Out[30]: True
In [31]: g44._supports_unit_fitting
Out[31]: False
```

pllim commented:
Any chance this is going into v2.0?

astrofrog commented:
@pllim - I really hope so! I'll take a look at the latest reviews now and see what I can do.

nden commented:
I started a proper code review some time ago but had no chance to finish it. I hope to do this next week.

Cadair commented:
@nden I would welcome your comments on how to add units to the projections module, I originally got involved with this because I wanted to have GWCS support units (specifically Longitude wrapping).

I started here: https://github.com/astropy/astropy/pull/4855/files#diff-e64cfea9de8722ddfc5104ec6ec2e26c

nden commented on astropy/modeling/core.py:
```diff
@@ -1546,7 +1846,7 @@ def _check_param_broadcast(self, params, max_ndim):
                                                     param_b, shape_b))
 
-    def _param_sets(self, raw=False):
+    def _param_sets(self, raw=False, units=False):
```
Adding units as a parameter somehow broke the shape of the returned array in the presence of units. For example

```
g1=models.Gaussian1D(10*u.A, 3*u.m, 1*u.m)
g1._param_sets(raw=True)
Out[20]: 
array([[ 10.],
       [  3.],
       [  1.]])

g1._param_sets(raw=True, units=True)
Out[19]: array([<Quantity [ 10.] A>, <Quantity [ 3.] m>, <Quantity [ 1.] m>], dtype=object)

```
This may seem insignificant but at some point it was important to have that shape. I can't tell right now without reading more whether this is still the case. However, I think we should try to be consistent.

astrofrog replied:
I think the shape is still equivalent, it's just that you have to nest the 1D quantities inside a 1D array rather than have a 2D array. We can't return 2D quantities with mismatching units.

nden commented:
@astrofrog Very nice. Thanks for your work. I left few minor comments inline. 

I am also wondering / hoping for this not to have a big performance hit on evaluation of models. A quick timeit on a Gaussian1D with an input array of size 2000000 shows 110 ms vs 95 ms but this will add up quickly for 2D and compound models. We may need to come up with a better scheme than checking the type of object in every function.

I just read the changes, will try running it now.

pllim commented on astropy/modeling/blackbody.py:
```diff
@@ -0,0 +1,95 @@
+        """
+
+        from ..constants import sigma_sb
+        from ..analytic_functions.blackbody import blackbody_nu
```
I think we can wait to move `blackbody_nu` over in a separate PR to deprecate `analytic_functions`.

astrofrog commented on astropy/modeling/core.py:
```diff
@@ -1036,6 +1097,99 @@ def has_user_bounding_box(self):
+        outputs_unit = {out: getattr(kwargs[out], 'unit', dimensionless_unscaled)
+                        for out in self.outputs if kwargs[out] is not None}
+
+        parameter_units = self._parameter_units_for_data_units(inputs_unit, outputs_unit)
```
This only gets called if ``_supports_unit_fitting`` returns True (which only works if ``_parameter_units_for_data_units`` exists)

astrofrog commented on astropy/modeling/blackbody.py:
```diff
@@ -0,0 +1,95 @@
+from __future__ import (absolute_import, unicode_literals, division,
```
Fixed

astrofrog commented on astropy/modeling/blackbody.py:
```diff
@@ -0,0 +1,95 @@
+    Parameters
+    ----------
+    temperature : :class:`~astropy.units.Quantity`
+        Blackbody temperature
```
Fixed

astrofrog commented:
@pllim - I've opened https://github.com/astropy/astropy/issues/6160 to remind us to deprecate and tidy up the blackbody stuff. This PR is becoming harder and harder to rebase so I'd like to merge it soon and make improvements like this in separate smaller PRs.

@nden - I've opened https://github.com/astropy/astropy/issues/6161 to remind myself to look at the performance implications. I think it will be easier to do this in a separate PR.

astrofrog commented:
This PR now includes docs. One last thing I want to do is to add tests for the Blackbody1D model.

astrofrog commented:
Rendered docs: http://astrofrog-debug.readthedocs.io/en/latest/modeling/units.html

bsipocz commented on docs/modeling/units.rst:
```diff
@@ -0,0 +1,279 @@
+On the other hand, if a parameter previously defined without units is given a
+Quantity with a unit, this works because it is unambiguous::
+
+    >>> g0 = Gaussian1D(mean=3)
```
nitpicking, but the name g0 is unintuitive when all the rest of the names are ordered, looking in the middle of the docs, I assumed it was already been used before g1

astrofrog replied:
Fixed, thanks!

pllim commented:
80 commits! 😱  I am okay with merging now and improve later as long as @nden approves. Also, deprecating `analytic_functions` already has an issue as #5780.

bsipocz commented:
@pllim - I would say 80 commits over 2 years by 4 authors for a net change of 1600+ lines is rather  good ratio. Keeping the history in cases like this is quite valuable I think.

astrofrog commented:
@pllim @nden - this is basically ready for final review. The rendered docs can be seen here:

http://astrofrog-debug.readthedocs.io/en/latest/modeling/units.html

It would be good if we could merge at some point within the next couple of days as I'll need to open a few other smaller PRs that will need to be ideally merged before feature freeze ❄️

nden commented:
Thanks for the hard work @astrofrog! :fireworks:
